### ISSUE TITLE
rk3399 fails to build: region PMUSRAM overflowed

### ISSUE THREAD

**Issue by vagrantc** [ARM-software/tf-issues#650]
Working on packaging arm-trusted-firmware in Debian.

Building arm-trusted-firmware from git commit cabe0a31801e99e7abb84d2114ded6bb56f3c71e, rk3399 fails to build with a recent Debian toolchain (gcc 8.2.0, binutils 3.31.1):

    Must remake target 'build/rk3399/debug/bl31/bl31.elf'.
  LD      build/rk3399/debug/bl31/bl31.elf
aarch64-linux-gnu-ld.bfd: build/rk3399/debug/bl31/bl31.elf section `.pmusram' will not fit in region `PMUSRAM'                         
aarch64-linux-gnu-ld.bfd: region `PMUSRAM' overflowed by 3928 bytes
make[2]: *** [Makefile:722: build/rk3399/debug/bl31/bl31.elf] Error 1                                                                  
make[2]: Leaving directory '/<<PKGBUILDDIR>>'

Full build log available at:

https://people.debian.org/~vagrant/atf/

Works fine building with the older toolchains (gcc 6, binutils 2.28) in Debian, but it's not possible to include arm-trusted-firmware in an older release; the newer toolchains are required to work for new uploads.

Thanks!

**Comment by antonio-nino-diaz-arm** [comment]
@jwerner-chromium

**Comment by antonio-nino-diaz-arm** [comment]
However, "overflowed by 3928 bytes" looks like there is some problem with the new version of the linker, not like GCC has become that worse in 2 major versions. There were only ~4KB of code for each CPU type (M0 and A72/53) to begin with.

**Comment by jwerner-chromium** [comment]
Agreed, this looks like a bug or serious regression, not just a natural change in code generation.

@rkchrome, can Rockchip people look into this?

**Comment by antonio-nino-diaz-arm** [comment]
It looks like @rkchrome is an old account and the correct one is @rockchip-linux.

**Comment by antonio-nino-diaz-arm** [comment]
Which version of the compilers are you using? I'm using the following compilers from here: https://developer.arm.com/open-source/gnu-toolchain/gnu-a/downloads
```
gcc-arm-8.2-2019.01-x86_64-arm-eabi
gcc-arm-8.2-2019.01-x86_64-aarch64-linux-gnu
```
And it seems to build correctly.

**Comment by maschinensehen** [comment]
I am getting a similar error on a current archlinux system with

aarch64-linux-gnu-gcc (GCC) 11.2.0
aarch64-linux-gnu-ld GNU ld (GNU Binutils) 2.38

LD      /data/arm-trusted-firmware/build/rk3399/release/bl31/bl31.elf
aarch64-linux-gnu-ld.bfd: /data/arm-trusted-firmware/build/rk3399/release/bl31/bl31.elf section `.pmusram' will not fit in region `PMUSRAM'
aarch64-linux-gnu-ld.bfd: .incbin_sram has exceeded its limit
aarch64-linux-gnu-ld.bfd: region `PMUSRAM' overflowed by 68376 bytes
make: *** [Makefile:1278: /data/arm-trusted-firmware/build/rk3399/release/bl31/bl31.elf] Error 1

**Comment by grandpaul** [comment]
I met the same issue. 
```
  CPP     bl31/bl31.ld.S
  LD      /home/paulliu/code/linaro-test-e5e925b80fd7/verify_gbl_on_rockpi4/tru\
sted-firmware-a/build/rk3399/release/bl31/bl31.elf
/usr/lib/gcc-cross/aarch64-linux-gnu/14/../../../../aarch64-linux-gnu/bin/ld: /\
home/paulliu/code/linaro-test-e5e925b80fd7/verify_gbl_on_rockpi4/trusted-firmwa\
re-a/build/rk3399/release/bl31/bl31.elf section `.pmusram' will not fit in regi\
on `PMUSRAM'
/usr/lib/gcc-cross/aarch64-linux-gnu/14/../../../../aarch64-linux-gnu/bin/ld: r\
egion `PMUSRAM' overflowed by 3928 bytes
collect2: error: ld returned 1 exit status
```
The build command is

`make CROSS_COMPILE=aarch64-linux-gnu- PLAT=rk3399 LOG_LEVEL=50`


I tried to bisect to find out the reason, and I found it is actually because the clang. 
The commit that causes the issue is.

```
6c2e5bf68955cdcce18fbd32697692e352e3c34c is the first bad commit
commit 6c2e5bf68955cdcce18fbd32697692e352e3c34c (HEAD)
Author: Boyan Karatotev <boyan.karatotev@arm.com>
Date:   Fri Apr 11 10:13:55 2025 +0100

    feat(build): use clang as a linker
    
    To support LTO, the gcc binary is used as a compiler, assembler, and
    linker. Do the same for clang and enable LTO builds with it as a side
    effect.
    
    This simplifies code quite a bit as the gcc/clang different is much
    smaller. Support for ld/lld (if overriden with LD) is maintained.
    
    This is a good time to convert tabs to spaces to conform to make's
    expectations on syntax.
    
    Change-Id: I6c529c1393f7e9e8046ed537f871fc3ad91d599a
    Signed-off-by: Boyan Karatotev <boyan.karatotev@arm.com>

 bl1/bl1.mk                             |  6 +--
 docs/getting_started/build-options.rst |  3 +-
 lib/cpus/cpu-ops.mk                    | 10 ++---
 make_helpers/build_macros.mk           | 13 +++---
 make_helpers/cflags.mk                 | 79 ++++++++++++++--------------------
 make_helpers/toolchain.mk              |  2 +-
 6 files changed, 48 insertions(+), 65 deletions(-)

```
I'm in Debian Trixie (13), the latest stable version of Debian. I'm wondering if this is caused by the version of clang.
clang version  in Debian Trixie is 19.1.7
